packages/dds/tree/src/simple-tree/arrayNode.ts

```diff
@@ -851,6 +865,22 @@ abstract class CustomArrayNodeBase<const T extends ImplicitAllowedTypes>
 			index,
 		);
 	}
+
+	public *values(): Generator<TreeNodeFromImplicitAllowedTypes<T>> {
+		const initialLastUpdatedStamp = this.#generationNumber;
+		let index = 0;
+		while (true) {
```

Comment:
You can have a for-loop that operates over two variables rather than one, if you prefer (and if the linter prefers - in the past it has complained about `while (true)` loops, though I'm not sure if that's still the case). IMO the for loop makes it more clear that the two variables are progressing together in lockstep.

```suggestion
		for (let i = 0, value = this.at(0); i < this.length && value !== undefined; i++, value = this.at(i)) {
```

Though, I am curious - under what circumstances is `value` undefined? If we only iterate up to `length`, and we know that length isn't going to fluctuate during the iteration, then aren't we guaranteed that `value` is always defined? Since we don't allow `undefined` in our arrays as legitimate elements.

Reply:
The reasoning behind the `while(true)` was that if we want to throw an error when we detect that the array has been edited during the iteration. 

For example (one of the test cases in this PR), if we used the for-loop above, and deleted a bunch of the array elements during the iteration, the `this.length` will have changed and `value` could be `undefined`. This unfortunately exits the for loop instead of making the check and throwing the error within the loop.

Would it be better to store the "original" array length into a local variable in the `values` method, and using this length in the for-loop while checking `value === undefined` and `i >= this.length` (which could be different from the original array length) and throw the error there instead of using the `while(true)` loop?

Reply:
I see, interesting. Yes, I suppose you could save the array length. I think you could also just be a little more careful about where you do the assert. The only time that the concurrent modification can happen is at the `yield` statement. So, instead of checking for the modification at the start of the loop, do it right after the `yield`. Then, you can write your loop however you want (e.g. as a for-loop, even using `this.length`) and you'll be fine.

One other thing this made me realize is that the generator code only runs lazily when the first item from the generator is requested. So, the baseline generation number that you save in `initialLastUpdatedStamp` is not necessarily the generation number when `values()` is called. Rather, it is the generation number when the iterator returned by `values()` is _first read_. So, a user could e.g. do this:

```ts
const values = myArray.values();
myArray.removeRange();
for (const v of values) {
  // ... No error
}
```
Basically, it allows modifications if the user has created, but not read from, the generator. Is that... okay? Seems maybe inconsistent to me? If you want to prevent that, you need to have values() not be a generator, but set `initialLastUpdatedStamp` and then pass it to another iterator (which could be a generator).

Reply:
Updated to for-loop for readability, added separate `*generateValues` method to pass in the correct `generationNumber`, and added test case for calling values, editing the array, and iterating